src/spec-configuration/containerFeaturesOrder.ts

```diff
@@ -3,133 +3,538 @@
+		// TODO
+		return false;
+	}
+
+	if (!featureSupportsDependencies(node.featureSet) || !featureSupportsDependencies(softDep.featureSet)) {
+		return false;
+	}
+
+	switch (nodeSourceInfo.type) {
+		case 'oci':
+			softDepSourceInfo = softDepSourceInfo as OCISourceInformation;
+			const nodeFeatureRef = nodeSourceInfo.featureRef;
+			const softDepFeatureRef = softDepSourceInfo.featureRef;
+			const softDepFeatureRefPrefix = `${softDepFeatureRef.registry}/${softDepFeatureRef.namespace}`;
+
+			return nodeFeatureRef.resource === softDepFeatureRef.resource || // Same resource
+				softDep.legacyIdAliases?.some(legacyId => `${softDepFeatureRefPrefix}/${legacyId}` === nodeFeatureRef.resource) || // Handle 'legacyIds'
+				false;
+
+		case 'direct-tarball':
+			throw new Error(`TODO: Should be supported but is unimplemented!`);
+		case 'file-path':
+			softDepSourceInfo = softDepSourceInfo as FilePathSourceInformation;
+			return nodeSourceInfo.resolvedFilePath === softDepSourceInfo.resolvedFilePath;
+		default:
+			throw new Error(`Feature dependencies are only supported in Features published (1) to an OCI registry, (2) as direct HTTPS tarball, or (3) as a local Feature.  Got type: '${nodeSourceInfo.type}'.`);
+	}
+}
+
+// If the two features are equal, return 0.
+// If the sorting algorithm should place A _before_ B, return negative number.
+// If the sorting algorithm should place A _after_  B, return positive number.
+function comparesTo(a: FNode, b: FNode): number {
+	const aSourceInfo = a.featureSet?.sourceInformation;
+	let bSourceInfo = b.featureSet?.sourceInformation; // Mutable only for type-casting.
+
+	if (!aSourceInfo || !bSourceInfo) {
+		// TODO: This indicates a bug - remove once confident this is not happening!
+		throw new Error(`ERR: Missing source information! (${a.userFeatureId} ?~ ${b.userFeatureId})`);
+	}
+
+	if (aSourceInfo.type !== bSourceInfo.type) {
+		return aSourceInfo.userFeatureId.localeCompare(bSourceInfo.userFeatureId);
+	}
+
+	switch (aSourceInfo.type) {
+		case 'oci':
+			bSourceInfo = bSourceInfo as OCISourceInformation;
+
+			const aResource = aSourceInfo.featureRef.resource;
+			const bResource = bSourceInfo.featureRef.resource;
+
+			const aDigest = aSourceInfo.manifestDigest;
+			const bDigest = bSourceInfo.manifestDigest;
+
+			const aOptions = JSON.stringify(a.options);
+			const bOptions = JSON.stringify(b.options);
+
+			const aCanonicalId = `${aResource}@${aDigest}`;
+			const bCanonicalId = `${bResource}@${bDigest}`;
+
+			if (aCanonicalId === bCanonicalId && aOptions === bOptions) {
+				// Equal!
+				return 0;
+			}
+
+			// Sort by resource name
+			if (aResource !== bResource) {
+				return aResource.localeCompare(bResource);
+			}
+
+			const aTag = aSourceInfo.featureRef.tag;
+			const bTag = bSourceInfo.featureRef.tag;
+			// Sort by tags (if both have tags)
+			// Eg: 1.9.9, 2.0.0, 2.0.1, 3, latest
+			if ((aTag && bTag) && (aTag !== bTag)) {
+				return aTag.localeCompare(bTag);
```

Comment:
It will result in a different sort since we're using the unresolved tags.  We could choose here not to compare the tags and instead compare solely on digest. The resolved digest will change unpredictably each release of the Feature. This feels more stable and intuitive to me, but at the end of the day the "stable order" here can be whatever we think is most intuitive.

Reply:
Having the latest version run last might have the advantage of that getting a chance at arranging things last, e.g., when previous versions have a bug.